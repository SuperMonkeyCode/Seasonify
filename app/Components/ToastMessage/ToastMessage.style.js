import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  container: {
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    height: 50,
    justifyContent: 'center',
  },

  // eslint-disable-next-line react-native/no-unused-styles
  neutral: {
    backgroundColor: '#0091ea',
  },
  // eslint-disable-next-line react-native/no-unused-styles
  success: {
    backgroundColor: '#00c853',
  },
  // eslint-disable-next-line react-native/no-unused-styles
  warning: {
    backgroundColor: '#ffab00',
  },
  // eslint-disable-next-line react-native/no-unused-styles
  error: {
    backgroundColor: '#d50000',
  },

  text: {
    color: 'white',
  },

  buttonText: {
    color: 'white',
    fontWeight: 'bold',
  },
});
