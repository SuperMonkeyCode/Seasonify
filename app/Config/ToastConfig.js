/* @flow */
import type { Toast, Toasts } from 'Types';

const baseDuration = 3000;

const defaults: Toast = {
  text: '',
  level: 'neutral',
  duration: baseDuration,
};

const toastConfig: Toasts = {
  neutral: {
    ...defaults,
  },

  success: {
    ...defaults,
    level: 'success',
  },

  warning: {
    ...defaults,
    level: 'warning',
    duration: baseDuration * 3,
  },

  error: {
    ...defaults,
    level: 'error',
    duration: baseDuration * 4,
  },
};

export default toastConfig;
