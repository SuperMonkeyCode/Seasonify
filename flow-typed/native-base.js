/* @flow */
declare module 'native-base' {
  declare var exports: any;
}
