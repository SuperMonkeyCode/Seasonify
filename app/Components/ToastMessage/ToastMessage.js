/* @flow */
import { Animated, StyleSheet } from 'react-native';
import { Button, Left, ListItem, Right, Text, View } from 'native-base';
import React from 'react';

import type { Message, Toast } from 'Types';
import ToastConfig from 'Config/ToastConfig';

import styles from './ToastMessage.style';

type DefaultProps = {
  useNativeDriver: boolean,
};

type Props = {
  message: Message,
  useNativeDriver: boolean,
  messageReset: Function,
  dispatch: Function,
};

type State = {
  modalShown: boolean,
  message?: ?Message,
  toastConfig: Toast,
  animationTiming: {
    duration: number,
    useNativeDriver: boolean,
  },
  height: number,
};

export default class ToastMessage extends React.Component<DefaultProps, Props, State> {
  static defaultProps = {
    useNativeDriver: true,
  };

  constructor(props: Props) {
    super(props);
    this.state.animationTiming.useNativeDriver = props.useNativeDriver;
  }

  state = {
    modalShown: false,
    message: undefined,
    toastConfig: ToastConfig.neutral,
    animationTiming: {
      duration: 350,
      useNativeDriver: true,
    },
    height: 0,
  };

  componentWillReceiveProps({ message }: Props) {
    if (message) {
      this.resetAnimation();

      /* istanbul ignore if. (A toast already displayed) */
      if (this.state.modalShown && this.closeTimeout) {
        this.removeCloseTimeout();
      }

      this.setState(
        {
          message,
          toastConfig: ToastConfig[message.level],
          height: this.heightFromStyles(),
        },
        this.openToast,
      );
    } else {
      this.closeToast();
    }
  }

  animation = new Animated.Value(0);
  closeTimeout = null;

  heightFromStyles = (): number => StyleSheet.flatten(styles.container).height;

  handleButtonCallback = () => {
    if (this.state.message && this.state.message.callback) {
      this.props.dispatch(this.state.message.callback);
    }
    this.closeToast();
  };

  handleToastPress = () => {
    this.closeToast();
  };

  resetAnimation() {
    this.animation = new Animated.Value(0);
  }

  showModal = () => {
    this.setState({ modalShown: true });
  };

  hideModal = () => {
    this.setState({ modalShown: false });
  };

  openToast = () => {
    this.showModal();
    this.programCloseToast();
    Animated.timing(this.animation, {
      ...this.state.animationTiming,
      toValue: 1,
    }).start();
  };

  programCloseToast = () => {
    this.closeTimeout = setTimeout(() => {
      this.props.messageReset();
    }, this.state.toastConfig.duration);
  };

  closeToast = () => {
    Animated.timing(this.animation, {
      ...this.state.animationTiming,
      toValue: 0,
    }).start(this.hideModal);

    this.removeCloseTimeout();
  };

  removeCloseTimeout = () => {
    clearTimeout(this.closeTimeout);
  };

  render() {
    const { message, toastConfig, height } = this.state;

    if (message && message.text) {
      const interpolated = this.animation.interpolate({
        inputRange: [0, 0.3, 1],
        outputRange: [0, height * -1 + 10, height * -1],
      });

      return (
        <View>
          <Animated.View
            style={[
              styles.container,
              styles[toastConfig.level],
              {
                transform: [{ translateY: interpolated }],
              },
            ]}
          >
            <ListItem onPress={this.handleToastPress}>
              <Left>
                <Text style={StyleSheet.flatten(styles.text)}>
                  {message.text}
                </Text>
              </Left>
              {message.button &&
                <Right>
                  <Button transparent onPress={this.handleButtonCallback}>
                    <Text style={StyleSheet.flatten(styles.buttonText)}>
                      {message.button}
                    </Text>
                  </Button>
                </Right>}
            </ListItem>
          </Animated.View>
        </View>
      );
    }

    return null;
  }
}
