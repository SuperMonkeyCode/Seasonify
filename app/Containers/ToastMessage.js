/* @flow */
import { Toast } from 'native-base';
import { compose, lifecycle } from 'recompose';
import { connect } from 'react-redux';

import type { Message } from 'Types';
import Messages from 'Config/Messages';
import ui from 'State/uiState';

export function componentWillReceiveProps({
  message,
  messageHide,
}: {
  message: Message,
  messageHide: Function,
}) {
  if (message) {
    const messageConfig = Messages[message.level];

    setTimeout(() => {
      messageHide();
    }, messageConfig.duration);
    Toast.show({ ...messageConfig, text: message.text });
  }
}

const enhance = compose(
  connect(
    state => ({
      message: ui.selectors.getMessage(state),
    }),
    {
      messageHide: ui.actions.messageHide,
    },
  ),
  lifecycle({
    componentWillReceiveProps,
  }),
);

export function ToastMessage() {
  return null;
}

export default enhance(ToastMessage);
