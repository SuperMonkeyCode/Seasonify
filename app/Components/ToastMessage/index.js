/* @flow */
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import uiState from 'State/uiState';

import ToastMessage from './ToastMessage';

export default connect(
  state => ({
    message: uiState.selectors.getMessage(state),
  }),
  dispatch => ({
    ...bindActionCreators({ messageReset: uiState.actions.messageReset }, dispatch),
    dispatch,
  }),

  // {
  //   messageReset: uiState.actions.messageReset,
  // },
)(ToastMessage);
