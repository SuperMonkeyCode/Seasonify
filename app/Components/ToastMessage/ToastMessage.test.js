import { Button } from 'native-base';
import React from 'react';

import ToastConfig from 'Config/ToastConfig';

import ToastMessage from './ToastMessage';

function setup(specificProps = {}) {
  const props = {
    message: null,
    useNativeDriver: false,
    messageReset: jest.fn(),
    dispatch: jest.fn(),
    ...specificProps,
  };
  const component = shallow(<ToastMessage {...props} />);
  return {
    component,
    props,
  };
}

describe('Rendering an empty message', () => {
  jest.useFakeTimers();
  const { props, component } = setup();

  it('should match initial state', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when receiving an empty message as a prop', () => {
    beforeEach(() => {
      component.setProps({
        ...props,
        message: null,
      });
    });

    it('should match hidden state', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when receiving message as a prop', () => {
    const fakeDuration = 5;
    beforeEach(() => {
      ToastConfig.success = { duration: fakeDuration };
      // setProps triggers componentWillReceiveProps too
      component.setProps({
        ...props,
        message: { level: 'success', text: 'Success message' },
      });
    });

    it('should match displayed state', () => {
      expect(component).toMatchSnapshot();
    });

    // TODO: Not working because Jest 20 doesn't call the setState's callback
    // cf. https://github.com/airbnb/enzyme/issues/953 and https://github.com/gitim/react-native-sortable-list/issues/50
    // it('should messageReset after a moment', () => {
    //   expect(setTimeout.mock.calls.length).toBe(1);
    //   expect(setTimeout.mock.calls[0][1]).toBe(fakeDuration);
    //   jest.runAllTimers();
    //   expect(props.messageReset).toBeCalled();
    // });
  });
});

describe('Rendering a warning message with an undo button', () => {
  const { props: initProps, component } = setup();
  const props = {
    ...initProps,
    message: {
      level: 'warning',
      text: 'Warning message',
      button: 'UNDO',
      callback: { type: 'UNDO' },
    },
  };
  component.setProps(props);

  it('should match', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when calling onPress on ChildButton', () => {
    const childButton = component.find(Button);
    childButton.props().onPress();

    it('calls dispatch with the action', () => {
      expect(props.dispatch).toBeCalledWith(props.message.callback);
    });
  });
});
