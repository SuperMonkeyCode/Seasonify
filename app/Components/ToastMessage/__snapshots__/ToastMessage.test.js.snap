// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a warning message with an undo button should match 1`] = `
<Styled(ViewNB)>
  <AnimatedComponent
    style={
      Array [
        Object {
          "height": 50,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "backgroundColor": "#ffab00",
        },
        Object {
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        },
      ]
    }
  >
    <Styled(ListItem)>
      <Styled(Left)>
        <Styled(Text)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Warning message
        </Styled(Text)>
      </Styled(Left)>
      <Styled(Right)>
        <Styled(Button)
          onPress={[Function]}
          transparent={true}
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontWeight": "bold",
              }
            }
          >
            UNDO
          </Styled(Text)>
        </Styled(Button)>
      </Styled(Right)>
    </Styled(ListItem)>
  </AnimatedComponent>
</Styled(ViewNB)>
`;

exports[`Rendering an empty message should match initial state 1`] = `null`;

exports[`Rendering an empty message when receiving message as a prop should match displayed state 1`] = `
<Styled(ViewNB)>
  <AnimatedComponent
    style={
      Array [
        Object {
          "height": 50,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        },
      ]
    }
  >
    <Styled(ListItem)>
      <Styled(Left)>
        <Styled(Text)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Success message
        </Styled(Text)>
      </Styled(Left)>
    </Styled(ListItem)>
  </AnimatedComponent>
</Styled(ViewNB)>
`;
